### PR TITLE
Add wrappers for `scipy.ndimage`

### LINKED ISSUE #2569: `scipy.ndimage` wrappers
Users would like to use https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.median_filter.html#scipy.ndimage.median_filter. It appears to be a good candidate for wrapping, similar to some other scipy subpackages. There are a number of very similar ones and we may be able to share a good amount of code between them.

One issue we came across, which currently appears to limit the scope of where our wrappers can be applied, is the limitation to "linspace" coords. Before we implement more wrappers we should figure out a better solution. Maybe we could simply warn/log if coords are not linspace? Or add an additional option to force handling of non-linspace?

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #2569.

Features:
- Works with variables as well as data arrays.
- `sigma` and `size` arguments passed as dicts of scalar variables instead of tuples.
- Can pass int/float instead of variable. In that case it refers to "positional" size. As discussed on Slack this may be too implicit, so we are considering to require an extra flag argument in that case (to ignore coords).

This is a draft for discussion.

TODO:
- [x] Add and require arg to explicitly ignore coords when passing sizes/sigma/other-params as plain indices instead of variables. -> see discussion below.
- [x] Docstrings
- [x] Unit tests
- [x] More functions

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
Here is an example if you want to try this yourself:
```python
import numpy as np
import scipp as sc
da = sc.DataArray(sc.array(dims=['x', 'y'], values=np.random.rand(100, 100)))
da.coords['x'] = sc.linspace('x', 0.0, 1.0, num=100, unit='mm')
da.coords['y'] = sc.linspace('y', 0.0, 5.0, num=100, unit='mm')

from scipp.ndimage import gaussian_filter, median_filter, maximum_filter

out = gaussian_filter(da, sigma={'x':sc.scalar(0.1, unit='mm'),
                                 'y':sc.scalar(0.2, unit='mm')})
#out = median_filter(da, size=3)
#out = median_filter(da, size=sc.scalar(0.3, unit='mm'))
#out = maximum_filter(da.data, size=3)
out.plot()
```

File: src/scipp/ndimage/__init__.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+from ..core import CoordError, DimensionError
+from ..core import empty_like, islinspace, ones
+
+
+def ndfilter(func: Callable) -> Callable:
+
+    @wraps(func)
+    def function(x: Union[Variable, DataArray], **kwargs) -> Union[Variable, DataArray]:
+        if 'output' in kwargs:
+            raise ValueError("The 'output' argument is not supported")
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:22]: I think this should be `TypeError`. This is what you usually get for specifying arguments that don't exist.

File: src/scipp/ndimage/__init__.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+
+@ndfilter
+def gaussian_filter(x: Union[Variable, DataArray],
+                    /,
+                    *,
+                    sigma,
+                    order=0,
+                    **kwargs) -> Union[Variable, DataArray]:
+    from scipy.ndimage import gaussian_filter
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:60]: Import at module level and make importing `scipp.ndimage` require that scipy is istsalled?

File: src/scipp/ndimage/__init__.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    if not islinspace(coord, dim).value:
+        raise CoordError(
+            f"Data points not regularly spaced along {dim}. To ignore this, "
+            "provide a plain value (int or float) instead of a scalar variable. "
+            "Note that this will correspond to plain positional indices/offsets.")
+    return (len(coord) - 1) * (index.to(unit=coord.unit) / (coord[-1] - coord[0])).value
+
+
+def _positional_index(x: Union[Variable, DataArray], index, name=None):
+    if isinstance(index, (int, Variable)):
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:45]: Missing `float`? (And in _delta_to_positional)

File: src/scipp/ndimage/__init__.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    order = order if isinstance(order, int) else [order[dim] for dim in x.dims]
+    out = empty_like(x)
+    gaussian_filter(x.values, sigma=sigma, order=order, output=out.values, **kwargs)
+    return out
+
+
+def _make_footprint(x: Union[Variable, DataArray], size, footprint) -> Variable:
+    if footprint is None:
+        size = _positional_index(x, size, name='size')
+        size = [int(s) for s in size]
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:71]: How does `scipy` handle floats in `size`?
Would it simplify things to pass a `size` to `scipy.ndimage.*_filter` and let it handle converting it to a footprint?

**Review by jl-wynen — COMMENTED** [review]
Looks like a good start according to our discussion.

**Comment by SimonHeybrock** [comment]
Here is an argument against requiring special parameters when specifying size/sigma as int/float:

`sc.mean` also ignores the coordinate and treats all points as equal. This is obviously wrong if you consider measuring the temperature over the course of, e.g., a day, with many measurements clustered in a specific hour. We can do little to nothing to protect form such cases.

**Comment by jl-wynen** [comment]
> sc.mean also ignores the coordinate and treats all points as equal

True. But it produces a single number and removes coords. So if we don't require an extra arg, we should at least remove coords.

**Comment by SimonHeybrock** [comment]
> > sc.mean also ignores the coordinate and treats all points as equal
> 
> True. But it produces a single number and removes coords. So if we don't require an extra arg, we should at least remove coords.

That is what I thought at first, but that would actually make it even worse!

- If we remove the coord, the consumer of the result has no way of telling that it is not evenly spaced.
- It can acquire different/conflicting coords by insertion into a dataset that has coords for the same times.

**Comment by SimonHeybrock** [comment]
> > sc.mean also ignores the coordinate and treats all points as equal
> 
> True. But it produces a single number and removes coords. So if we don't require an extra arg, we should at least remove coords.

Furthermore, it would not make sense to drop the coords if they are linspace... with the effect of making the result of the operation unpredictable --- in some cases you'd get coords, in others you would not, and that could happen simply from different accumulation of rounding errors.

**Comment by SimonHeybrock** [comment]
Having slept over this a couple more times, I think we should *not* require linspace coords (or add flags/tolerance-args) when integer size/offset-related arguments are given. We need to make a clear distinction:

1. *Requirements* for an operation.
2. What *we think* is *correct*.

Examples will be necessary to clarify what I mean:

- For an addition operation it is a *requirement* that we units match, we cannot at 1 Kelvin to 1 meter.
- For binary operations it is a *requirement* that dims and coords match, since we do not automatically align / fill with default values.
- We *think* computing the `sum()` of a variable with `unit='Kelvin'` is (probably) wrong. It is however, not a *requirement* for the operation to work. There are definitely "edge" cases where this *is* what the user wants to do. scipp should not get in the way.
- We *think* it is probably "wrong" to have masks that are not `dtype='bool'` or that have a unit. But it is not a requirement for `DataArray`. It *becomes a requirement* when we perform a binary operation that needs to operate on the masks.
- For `gaussian_filter` with `sigma=12*sc.Unit('mm')` it is a *requirement* that the coords are "linspace" since we do not have an advanced algorithm to deal with changing sizes and therefore simply need to convert to an underlying integer offset.
- For `gaussian_filter` with `sigma=5` there is no such requirement. We can (mathematically) apply the operation. Do we *think* that it is not correct in a number of cases? Yes. But there are many cases where this is what the user wants to do.

In summary, we must (and are) preventing conceptually impossible operations, but we must acknowledge that it is impossible to protect from all possible user errors. Attempts to put extra safeguards/checks in place to prevent "things we thing are probably wrong" are necessarily incomplete, complicate the code, and lead to a worsened user experience. This has been observed in the past, and scipp has largely avoided making the same problematic design choice.

File: src/scipp/ndimage/__init__.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def _delta_to_positional(x: Union[Variable, DataArray], dim, index):
+    if isinstance(index, int):
+        return index
+    coord = x.coords[dim]
+    if not islinspace(coord, dim).value:
+        raise CoordError(
+            f"Data points not regularly spaced along {dim}. To ignore this, "
+            "provide a plain value (int or float) instead of a scalar variable. "
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:39]: Good catch. This is a mixup between `gaussian_filter` (where `sigma` can be float) and the others (where size must be an integer). Will fix.

File: src/scipp/ndimage/__init__.py (hunk 4 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:71]: I don't think it accepts floats in size.

File: src/scipp/ndimage/__init__.py (hunk 3 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:45]: Yes, see also Neil's comment above.

Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py]: Please check in the latest push how I fixed this.

File: src/scipp/ndimage/__init__.py [hunk 6]
```diff
@@ -31,25 +31,27 @@ def function(x: Union[Variable, DataArray], **kwargs) -> Union[Variable, DataArr
...
+            "indices/offsets.")
+    pos = (len(coord) - 1) * (index.to(unit=coord.unit) / (coord[-1] - coord[0])).value
+    return dtype(pos)
 
 
-def _positional_index(x: Union[Variable, DataArray], index, name=None):
-    if isinstance(index, (int, Variable)):
-        return [_delta_to_positional(x, dim, index) for dim in x.dims]
+def _positional_index(x: Union[Variable, DataArray], index, name=None, dtype=int):
+    if not hasattr(index, '__iter__'):
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:49]: This check looks strange. What cases does it handle?

Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:49]: The code later handles anything that is a mapping. I could just check for `not isinstance(index, dict)` and hope no one tries to use other mappings?

Comment (jl-wynen) [reply src/scipp/ndimage/__init__.py:49]: What if in the future, `Variable` has an iterator? I think a more concrete check would be safer. It could also be `hasattr(index, 'keys')`. This is how `dict.update` identifies dict-like types.

File: src/scipp/ndimage/__init__.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    if not isinstance(index, dict):
+        return [_delta_to_positional(x, dim, index, dtype=dtype) for dim in x.dims]
+    if set(index) != set(x.dims):
+        raise ValueError(f"Data has dims={x.dims} but input argument '{name}' provides "
+                         f"values for {tuple(index)}")
+    return [_delta_to_positional(x, dim, index[dim], dtype=dtype) for dim in x.dims]
+
+
+@_ndfilter
+def gaussian_filter(x: Union[Variable, DataArray],
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:59]: Ideally, this would use `VariableLikeType` (or some other `TypeVar` restricted to `Variable` and `DataArray`) and the same in the return type. This would tell type checkers that the returned type is the same as the input.
This does however make the documentation harder to read which is why I put `x: scipp.typing.VariableLike` in the docstrings.

File: src/scipp/ndimage/__init__.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+                    /,
+                    *,
+                    sigma,
+                    order=0,
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:63]: Type annotations?

File: src/scipp/ndimage/__init__.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+                    **kwargs) -> Union[Variable, DataArray]:
+    """
+    Multidimensional Gaussian filter.
+
+    This is a wrapper around :py:func:`scipy.ndimage.gaussian_filter`. See there for
+    full argument description. There are two key differences:
+
+    - This wrapper uses explicit dimension labels in the ``sigma`` and ``order``
+      arguments. For example, instead of ``sigma=[4, 6]`` use
+      ``sigma={{'time':4, 'space':6}}``
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:73]: Why double braces?

File: src/scipp/ndimage/__init__.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+      (with appropriate dimension labels for the data).
+    - Coordinate values can be used (and should be preferred) for ``sigma``. For
+      example, instead of ``sigma=[4, 6]`` use
+      ``sigma={{'time':sc.scalar(5.0, unit='ms'), 'space':sc.scalar(1.2, unit='mm')}}``.
+      In this case it is required that the corresponding coordinates exist and form a
+      "linspace", i.e., are evenly spaced.
+
+    Warning
+    -------
+    When ``sigma`` is an integer or an mapping to integers coordinate values are
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:83]:
```suggestion
    When ``sigma`` is an integer or a mapping to integers, coordinate values are
```

File: src/scipp/ndimage/__init__.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+    ignored. That is, the filter is applied even if the data points are not evenly
+    spaced. The resulting filtered data may thus have no meaningful interpretation.
+
+    Parameters
+    ----------
+    x:
+        Input variable or data array.
+    sigma:
+        Standard deviation for Gaussian kernel. The standard deviations of the Gaussian
+        filter are given as a mapping from dimension labels to numbers, or as a single
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:93]:
```suggestion
        filter are given as a mapping from dimension labels to numbers or variables, or as a single
```
?

File: src/scipp/ndimage/__init__.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+
+def _delta_to_positional(x: Union[Variable, DataArray], dim, index, dtype):
+    if not isinstance(index, Variable):
+        return index
+    coord = x.coords[dim]
+    if not islinspace(coord, dim).value:
+        raise CoordError(
+            f"Data points not regularly spaced along {dim}. To ignore this, "
+            f"provide a plain value (convertible to {dtype.__name__}) instead of a "
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:42]: I don't think this works:
```.py
>>> sc.DType.float64.__name__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'scipp._scipp.core.DType' object has no attribute '__name__'
```
Why did you not use `{dtype}`?

Or is `dtype` only allowed to be `int` and `float`? (Looking at `dtype(pos)` below)

File: src/scipp/ndimage/__init__.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+
+      >>> filtered = gaussian_filter(da, sigma=sc.scalar(1.2, unit='mm'))
+
+    With different sigma for different dimensions:
+
+      >>> filtered = gaussian_filter(da, sigma={'x':sc.scalar(1.2, unit='mm'), \
+                                                'y':sc.scalar(3.3, unit='mm')})
+    """
+    sigma = _positional_index(x, sigma, name='sigma', dtype=float)
+    order = order if isinstance(order, int) else [order[dim] for dim in x.dims]
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:121]: This should have the same dimension check as `_positional_index`.

File: src/scipp/ndimage/__init__.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+    return footprint
+
+
+def _make_footprint_filter(name, example=True, extra_args=''):
+
+    def footprint_filter(x: Union[Variable, DataArray],
+                         /,
+                         *,
+                         size: Union[int, Variable, Dict[str, Union[int,
+                                                                    Variable]]] = None,
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:150]:
```suggestion
                         size: Optional[Union[int, Variable, Dict[str, Union[int,
                                                                    Variable]]]] = None,
```

File: src/scipp/ndimage/__init__.py [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+      ``size={{'time':4, 'space':6}}`` (with appropriate dimension labels for the data).
+    - Coordinate values can be used (and should be preferred) for ``size`` and
+      ``origin``. For example, instead of ``size=[4, 6]`` use
+      ``size={{'time':sc.scalar(5.0, unit='ms'), 'space':sc.scalar(1.2, unit='mm')}}``.
+      In this case it is required that the corresponding coordinates exist and form a
+      "linspace", i.e., are evenly spaced.
+
+    Warning
+    -------
+    When ``size`` is an integer or an mapping to integers or when ``footprint`` is
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:186]:
```suggestion
    When ``size`` is an integer or a mapping to integers or when ``footprint`` is
```

File: src/scipp/ndimage/__init__.py [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+    data points are not evenly spaced. The resulting filtered data may thus have no
+    meaningful interpretation.
+
+    Parameters
+    ----------
+    x:
+        Input variable or data array.
+    size:
+        Integer or scalar variable or mapping from dimension labels to integers or
+        scalar variables. Defines the footprint (see below).
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:197]: Don't need to describe the type here, the type hint should take care of that.

File: src/scipp/ndimage/__init__.py [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+    sigma:
+        Standard deviation for Gaussian kernel. The standard deviations of the Gaussian
+        filter are given as a mapping from dimension labels to numbers, or as a single
+        number, in which case it is equal for all axes.
+    order:
+        The order of the filter along each dimension, given as mapping from dimension
+        labels to integers, or as a single integer. An order of 0 corresponds to
+        convolution with a Gaussian kernel. A positive order corresponds to convolution
+        with that derivative of a Gaussian.
+
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:100]: Note that you need to put a `Returns` section to properly show the return type in the docs.

File: src/scipp/ndimage/__init__.py [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+            raise ValueError("Provide either 'size' or 'footprint', not both.")
+        if set(footprint.dims) != set(x.dims):
+            raise DimensionError(
+                f"Dimensions {footprint.dims} must match data dimensions {x.dim}")
+    return footprint
+
+
+def _make_footprint_filter(name, example=True, extra_args=''):
+
+    def footprint_filter(x: Union[Variable, DataArray],
```
Comment (jl-wynen) [inline src/scipp/ndimage/__init__.py:146]: Missing `@_ndfilter`?

File: tests/ndimage/footprint_filter_test.py [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+from scipp.ndimage import (generic_filter, maximum_filter, median_filter,
+                           minimum_filter, percentile_filter, rank_filter)
+import pytest
+
+filters = (generic_filter, maximum_filter, median_filter, minimum_filter,
+           percentile_filter, rank_filter)
+
+
+def make_histogram2d():
+    da = sc.DataArray(sc.array(dims=['x', 'y'], values=np.random.rand(100, 100)))
```
Comment (jl-wynen) [inline tests/ndimage/footprint_filter_test.py:15]: Not seeding the rng might make the tests flaky and unreproducible. Unless you print `da` in its entirety.

File: tests/ndimage/footprint_filter_test.py [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+    with pytest.raises(TypeError):
+        filter_func(da, size=[2, 3])
+
+
+def test_raises_TypeError_if_size_is_not_integral(filter_func):
+    da = make_histogram2d()
+    with pytest.raises(TypeError):
+        filter_func(da, size=1.5)
+    with pytest.raises(TypeError):
+        filter_func(da, size=[2.5, 3.5])
```
Comment (jl-wynen) [inline tests/ndimage/footprint_filter_test.py:79]: Should this use a dict instead? Otherwise it fails because of the list, not because of floats. (Same below)

File: tests/ndimage/footprint_filter_test.py [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+import scipp as sc
+import scipy.ndimage
+from scipp.ndimage import (generic_filter, maximum_filter, median_filter,
+                           minimum_filter, percentile_filter, rank_filter)
+import pytest
+
+filters = (generic_filter, maximum_filter, median_filter, minimum_filter,
+           percentile_filter, rank_filter)
+
+
```
Comment (jl-wynen) [inline tests/ndimage/footprint_filter_test.py:13]: Very thorough :+1:

File: src/scipp/ndimage/__init__.py (hunk 9 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:73]: This is an fstring, double to escape them.

File: src/scipp/ndimage/__init__.py (hunk 12 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:42]: `dtype` is only `int` or `float`, yes. `{dtype}` renders to something like `"<class 'int'>"`, so it is less user readable.

File: src/scipp/ndimage/__init__.py (hunk 9 above)
Comment (jl-wynen) [reply src/scipp/ndimage/__init__.py:73]: This one isn't. Only the one in `footprint_filter` is.

File: src/scipp/ndimage/__init__.py (hunk 12 above)
Comment (jl-wynen) [reply src/scipp/ndimage/__init__.py:42]: ok

File: src/scipp/ndimage/__init__.py (hunk 9 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:73]: 🤦

File: src/scipp/ndimage/__init__.py (hunk 11 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:93]: 👍

File: tests/ndimage/footprint_filter_test.py (hunk 20 above)
Comment (SimonHeybrock) [reply tests/ndimage/footprint_filter_test.py:79]: Good catch!

File: tests/ndimage/footprint_filter_test.py (hunk 21 above)
Comment (SimonHeybrock) [reply tests/ndimage/footprint_filter_test.py:13]: Actually I am cheating, since not all tests are using the fixture that gets all filters.

File: tests/ndimage/footprint_filter_test.py (hunk 19 above)
Comment (SimonHeybrock) [reply tests/ndimage/footprint_filter_test.py:15]: Does that apply in this case? We are not checking concrete values, only compare to scipy.

File: src/scipp/ndimage/__init__.py (hunk 13 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py:121]: ... so there are some missing tests ;)

Good catch!

File: src/scipp/ndimage/__init__.py (hunk 16 above)
Comment (jl-wynen) [reply src/scipp/ndimage/__init__.py:197]: True. But that could be expressed more succinctly if you want to.

File: src/scipp/ndimage/__init__.py (hunk 18 above)
Comment (jl-wynen) [reply src/scipp/ndimage/__init__.py:146]: I see, I missed that.

File: tests/ndimage/footprint_filter_test.py (hunk 19 above)
Comment (jl-wynen) [reply tests/ndimage/footprint_filter_test.py:15]: Does it affect conversion of variables in arguments to integers? I don't think so? So you are right, it shouldn't matter.

File: tests/ndimage/footprint_filter_test.py (hunk 21 above)
Comment (jl-wynen) [reply tests/ndimage/footprint_filter_test.py:13]: I meant in general. But apparently we both missed a case

File: src/scipp/ndimage/__init__.py (hunk 7 above)
Comment (SimonHeybrock) [reply src/scipp/ndimage/__init__.py]: See update.

File: tests/ndimage/footprint_filter_test.py (hunk 20 above)
Comment (jl-wynen) [reply tests/ndimage/footprint_filter_test.py:79]: You changed the test below but not this one. Was that on purpose?

Comment (SimonHeybrock) [reply tests/ndimage/footprint_filter_test.py]: No. Done now.

**Review by jl-wynen — APPROVED** [review]
